membership-attribute-service/app/controllers/AccountController.scala

```diff
@@ -35,34 +37,38 @@ class AccountController extends LazyLogging {
+      user <- EitherT(Future.successful(authenticationService.userId \/> "no identity cookie for user"))
+      sfUser <- EitherT(tp.contactRepo.get(user).map(_ \/> s"couldn't read contact from SF for $user (TODO check the separate ERROR to find out why)"))
+      subscription <- EitherT(tp.subService.current[P](sfUser).map(_.headOption).map (_ \/> s"no current subscriptions for the sfUser $sfUser"))
+      stripeCardToken <- EitherT(Future.successful(updateForm.bindFromRequest().value \/> "no card token submitted with request"))
+      updateResult <- EitherT(tp.paymentService.setPaymentCardWithStripeToken(subscription.accountId, stripeCardToken).map(_ \/> "something missing when try to zuora payment card"))
     } yield updateResult match {
       case success: CardUpdateSuccess => Ok(Json.toJson(success))
       case failure: CardUpdateFailure => Forbidden(Json.toJson(failure))
-    }).run.map(_.getOrElse(notFound))
+    }).run.map {
+      case -\/(message) =>
+        logger.warn(s"didn't update card, $message")
+        notFound
+      case \/-(result) => result
+    }
   }
 
   def paymentDetails[P <: SubscriptionPlan.Paid : SubPlanReads, F <: SubscriptionPlan.Free : SubPlanReads] = mmaAction.async { implicit request =>
     val maybeUserId = authenticationService.userId
     logger.info(s"Attempting to retrieve payment details for identity user: $maybeUserId")
     (for {
-      user <- OptionT(Future.successful(maybeUserId))
-      contact <- OptionT(request.touchpoint.contactRepo.get(user))
-      sub <- OptionT(request.touchpoint.subService.either[F, P](contact))
-      details <- OptionT(request.touchpoint.paymentService.paymentDetails(sub).map[Option[PaymentDetails]](Some(_)))
-    } yield (contact, details).toResult).run.map { maybeResult => maybeResult match {
-        case Some(result) => {
-          logger.info(s"Successfully retrieved payment details result for identity user: $maybeUserId")
-          result
-        }
-        case None => {
-          logger.info(s"Unable to retrieve payment details result for identity user $maybeUserId")
-          Ok(Json.obj())
-        }
+      user <- EitherT(Future.successful(maybeUserId \/> "no identity cookie for user"))
+      contact <- EitherT(request.touchpoint.contactRepo.get(user).map(_ \/> s"couldn't read contact from SF for $user (TODO check the separate ERROR to find out why)"))
+      sub <- EitherT(request.touchpoint.subService.either[F, P](contact).map(_ \/> s"couldn't read sub from zuora for crmId ${contact.salesforceAccountId} (TODO check the separate WARN to find out why)"))
+      details <- EitherT(request.touchpoint.paymentService.paymentDetails(sub).map[\/[String, PaymentDetails]](\/.right))
+    } yield (contact, details).toResult).run.map {
+      case \/-(result) => {
+        logger.info(s"Successfully retrieved payment details result for identity user: $maybeUserId")
+        result
+      }
+      case -\/(message) => {
```

Comment:
this also covers the case is they somehow turn up at the manage page without being logged in.